tests/utils/stl/compiler.py

```diff
@@ -0,0 +1,127 @@
+#===----------------------------------------------------------------------===##
+#
+# Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+#===----------------------------------------------------------------------===##
+
+import os
+
+import stl.util
+
+
+class CXXCompiler(object):
+    CM_Default = 0
+    CM_PreProcess = 1
+    CM_Compile = 2
+    CM_Link = 3
+
+    def __init__(self, path, flags=None, compile_flags=None,
+                 link_flags=None, compile_env=None):
+        self.path = path
+        if path:
+            self.name = os.path.basename(path).split('.')[0]
+        else:
+            self.name = None
+
+        self.compile_flags = compile_flags or []
+        self.flags = flags or []
+        self.link_flags = link_flags or []
+
+        if compile_env is not None:
+            self.compile_env = dict(compile_env)
+        else:
+            self.compile_env = None
+
+    def _basicCmd(self, source_files, out, mode=CM_Default, flags=[]):
+        cmd = []
+        cmd += [self.path]
+        if mode == self.CM_PreProcess:
+            if out is not None:
+                cmd += ['/P', '/Fi' + out]
+            else:
+                cmd += ['/EP']
+        elif mode == self.CM_Compile:
+            cmd += ['/c']
+            if out is not None:
+                cmd += ['/Fo' + out]
+        elif out is not None:
+            cmd += ['/Fe' + out]
+
+        if mode != self.CM_Link:
+            cmd += self.compile_flags
+        cmd += self.flags
+        cmd += flags
+        if isinstance(source_files, list):
+            cmd += source_files
+        elif isinstance(source_files, str):
+            cmd += [source_files]
+        else:
+            raise TypeError('source_files must be a string or list')
+
+        if mode != self.CM_PreProcess and mode != self.CM_Compile:

```

Comment:
I'd prefer inverting this, as that seems to be the right thing if you ever add more modes (also I think it's clearer